SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityPendingTaskQueueProcessor.java

```diff
@@ -0,0 +1,419 @@
+  private final ExecutorService lockWaitExecutor;
+  private final ExecutorService taskLaunchExecutor;
+  private final ExecutorService queuePollExecutor;
+  private final AtomicBoolean running;
+  private final Resources defaultResources;
+  private final Resources defaultCustomExecutorResources;
+
+  private Future<?> lockFuture = null;
+
+  @Inject
+  public SingularityPendingTaskQueueProcessor(
+    SingularityOfferScoring offerScoring,
+    SingularityMesosOfferManager offerManager,
+    UsageManager usageManager,
+    SingularityConfiguration configuration,
+    TaskRequestManager taskRequestManager,
+    TaskManager taskManager,
+    RequestManager requestManager,
+    PriorityManager priorityManager,
+    SingularityManagedThreadPoolFactory threadPoolFactory,
+    SingularitySchedulerMetrics metrics,
+    SingularitySchedulerLock lock,
+    SingularityMesosSchedulerClient schedulerClient,
+    SingularityMesosTaskBuilder mesosTaskBuilder,
+    SingularityTaskSizeOptimizer taskSizeOptimizer
+  ) {
+    this.offerScoring = offerScoring;
+    this.offerManager = offerManager;
+    this.usageManager = usageManager;
+    this.configuration = configuration;
+    this.mesosConfiguration = configuration.getMesosConfiguration();
+    this.taskRequestManager = taskRequestManager;
+    this.taskManager = taskManager;
+    this.requestManager = requestManager;
+    this.priorityManager = priorityManager;
+    this.preLockQueue =
+      new PriorityBlockingQueue<>(
+        11,
+        (a, b) -> Doubles.compare(getWeightedPriority(a), getWeightedPriority(b))
+      );
+    this.queuePollExecutor = threadPoolFactory.getSingleThreaded("pending-task-queue");
+    this.lockWaitExecutor =
+      threadPoolFactory.get("pending-task-lock-wait", PARALLEL_LOCK_WAIT);
+    this.taskLaunchExecutor = threadPoolFactory.getSingleThreaded("task-launch");
+    this.running = new AtomicBoolean(false);
+    this.metrics = metrics;
+    this.lock = lock;
+    this.schedulerClient = schedulerClient;
+    this.mesosTaskBuilder = mesosTaskBuilder;
+    this.taskSizeOptimizer = taskSizeOptimizer;
+    this.lockStarts = new ConcurrentHashMap<>();
+    this.handled = new HashSet<>();
+    this.defaultResources =
+      new Resources(
+        mesosConfiguration.getDefaultCpus(),
+        mesosConfiguration.getDefaultMemory(),
+        0,
+        mesosConfiguration.getDefaultDisk()
+      );
+    this.defaultCustomExecutorResources =
+      new Resources(
+        configuration.getCustomExecutorConfiguration().getNumCpus(),
+        configuration.getCustomExecutorConfiguration().getMemoryMb(),
+        0,
+        configuration.getCustomExecutorConfiguration().getDiskMb()
+      );
+  }
+
+  public void start() {
+    if (lockFuture != null) {
+      lockFuture.cancel(true);
+    }
+    running.set(true);
+    lockFuture = queuePollExecutor.submit(this::runLockWait);
+  }
+
+  public void addPendingTaskIfNotInQueue(SingularityPendingTaskId pendingTask) {
```

Comment:
Does the scheduler need to support deleting/tombstoning tasks, like if a new deploy overwrites a pending deploy?